File: csharp/src/TimeSeries.cs
namespace NachoDB
{
    using System;
#if USE_GENERICS
    using System.Collections.Generic;
#else
    using System.Collections;
#endif

    /// <summary>
    /// Interface for timeseries element.
    /// You should derive your time series element from this class
    /// and implement Time getter method.
    /// </summary>
    public interface TimeSeriesTick 
    { 
        /// <summary>
        /// Get time series element timestamp (100 nanoseconds)
        /// </summary>
        long Time {get;}
    }

#if !USE_GENERICS
    /// <summary>
    /// Abstract base class for time series block.
    /// Progammer has to define its own block class derived from this class
    /// containign array of time series elements and providing accessors to the array elements 
    /// and Ticks getter method to access this whole array.
    /// </summary>
    public abstract class TimeSeriesBlock : Persistent 
    { 
        public long timestamp;
        public int  used;

        /// <summary>
        /// Get time series elements stored in this block.
        /// Returns preallocated array of time series element. Only <code>used</code>
        /// items of this array actually contains time series elements.
        /// </summary>
        public abstract Array Ticks{get;}

        /// <summary>
        /// Array elements accessor. 
        /// </summary>
        public abstract TimeSeriesTick this[int i] {get; set;}
    }
#endif

    /// <summary>
    /// <p>
    /// Time series interface. Time series class is used for efficient
    /// handling of time series data. Ussually time series contains a very large number
    /// if relatively small elements which are ussually acessed in sucessive order. 
    /// To avoid overhead of loading from the disk each particular time series element, 
    /// this class group several subsequent time series elements together and store them 
    /// as single object (block).
    /// </p><p> 
    /// As far as C# currently has no templates and
    /// Perst need to know format of block class, it is responsibity of prgorammer
    /// to create block implementation derived from TimeSeriesBlock class
    /// and containing array of time series elements. Size of this array specifies
    /// the size of the block.
    /// </p>
    /// </summary>
#if USE_GENERICS
    public interface TimeSeries<T> : IPersistent, IResource, ICollection<T> where T:TimeSeriesTick
#else
    public interface TimeSeries : IPersistent, IResource, ICollection 
#endif
    {    
#if !USE_GENERICS
        /// <summary>
        /// Add new tick to time series
        /// </summary>
        /// <param name="tick">new time series element</param>
        void Add(TimeSeriesTick tick);    
#endif

        /// <summary>
        /// Get forward iterator for time series elements belonging to the specified range
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <param name="till">inclusive time of the ending of interval</param>
        /// <returns>forward iterator within specified range</returns>
#if USE_GENERICS
        IEnumerator<T> GetEnumerator(DateTime from, DateTime till);
#else
        IEnumerator GetEnumerator(DateTime from, DateTime till);
#endif

        /// <summary>
        /// Get iterator through all time series elements
        /// </summary>
        /// <param name="order">direction of iteration</param>
        /// <returns>iterator in specified direction</returns>
#if USE_GENERICS
        IEnumerator<T> GetEnumerator(IterationOrder order);
#else
        IEnumerator GetEnumerator(IterationOrder order);
#endif

        /// <summary>
        /// Get forward iterator for time series elements belonging to the specified range
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <param name="till">inclusive time of the ending of interval</param>
        /// <param name="order">direction of iteration</param>
        /// <returns>iterator within specified range in specified direction</returns>
#if USE_GENERICS
        IEnumerator<T> GetEnumerator(DateTime from, DateTime till, IterationOrder order);
#else
        IEnumerator GetEnumerator(DateTime from, DateTime till, IterationOrder order);
#endif

        /// <summary>
        /// Get forward iterator for time series elements belonging to the specified range
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <param name="till">inclusive time of the ending of interval</param>
        /// <returns>forward iterator within specified range</returns>
#if USE_GENERICS
        IEnumerable<T> Range(DateTime from, DateTime till);
#else
        IEnumerable Range(DateTime from, DateTime till);
#endif

        /// <summary>
        /// Get iterator through all time series elements
        /// </summary>
        /// <param name="order">direction of iteration</param>
        /// <returns>iterator in specified direction</returns>
#if USE_GENERICS
        IEnumerable<T> Range(IterationOrder order);
#else
        IEnumerable Range(IterationOrder order);
#endif

        /// <summary>
        /// Get forward iterator for time series elements belonging to the specified range
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <param name="till">inclusive time of the ending of interval</param>
        /// <param name="order">direction of iteration</param>
        /// <returns>iterator within specified range in specified direction</returns>
#if USE_GENERICS
        IEnumerable<T> Range(DateTime from, DateTime till, IterationOrder order);
#else
        IEnumerable Range(DateTime from, DateTime till, IterationOrder order);
#endif

        /// <summary>
        /// Get forward iterator for time series elements with timestamp greater or equal than specified
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <returns>forward iterator</returns>
#if USE_GENERICS
        IEnumerable<T> From(DateTime from);
#else
        IEnumerable From(DateTime from);
#endif

        /// <summary>
        /// Get backward iterator for time series elements with timestamp less or equal than specified
        /// </summary>
        /// <param name="till">inclusive time of the eding of interval</param>
        /// <returns>backward iterator</returns>
#if USE_GENERICS
        IEnumerable<T> Till(DateTime till);
#else
        IEnumerable Till(DateTime till);
#endif

        /// <summary>
        /// Get backward iterator for time series elements 
        /// </summary>
        /// <returns>backward iterator</returns>
#if USE_GENERICS
        IEnumerable<T> Reverse();
#else
        IEnumerable Reverse();
#endif

        /// <summary>
        /// Get timestamp of first time series element
        /// </summary>
        /// <exception cref="NachoDB.StorageError">StorageError(StorageError.ErrorClass.KEY_NOT_FOUND) if time series is empy</exception>
        DateTime FirstTime {get;}

        /// <summary>
        /// Get timestamp of last time series element
        /// </summary>
        /// <exception cref="NachoDB.StorageError">StorageError(StorageError.ErrorClass.KEY_NOT_FOUND) if time series is empy</exception>
        DateTime LastTime {get;}

        /// <summary> 
        /// Get tick for specified data
        /// </summary>
        /// <param name="timestamp">time series element timestamp</param>
        /// <exception cref="NachoDB.StorageError">StorageError(StorageError.ErrorClass.KEY_NOT_FOUND) if no element with such timestamp exists</exception>
#if USE_GENERICS
        T this[DateTime timestamp] 
#else
        TimeSeriesTick this[DateTime timestamp] 
#endif
        {
            get;
        }
    
        /// <summary>
        /// Check if data is available in time series for the specified time
        /// </summary>
        /// <param name="timestamp">time series element timestamp</param>
        /// <returns><code>true</code> if there is element in time series with such timestamp, 
        /// <code>false</code> otherwise</returns>
        bool Contains(DateTime timestamp);

        /// <summary>
        /// Remove time series elements belonging to the specified range
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <param name="till">inclusive time of the ending of interval</param>
        /// <returns>number of removed elements</returns>
        int Remove(DateTime from, DateTime till);

        /// <summary>
        /// Remove time series elements with timestamp greater or equal then specified
        /// </summary>
        /// <param name="from">inclusive time of the begging of interval</param>
        /// <returns>number of removed elements</returns>
        int RemoveFrom(DateTime from);

        /// <summary>
        /// Remove time series elements with timestamp less or equal then specified
        /// </summary>
        /// <param name="till">inclusive time of the ending of interval</param>
        /// <returns>number of removed elements</returns>
        int RemoveTill(DateTime till);

        /// <summary>
        /// Remove all time series elements
        /// </summary>
        /// <returns>number of removed elements</returns>
        int RemoveAll();
    }
}

File: java/src15/org/nachodb/TimeSeries.java
package org.nachodb;

import java.util.*;

/**
 * <p>
 * Time series interface. Time series class is used for efficient
 * handling of time series data. Ussually time series contains a very large number
 * if relatively small elements which are ussually acessed in sucessive order. 
 * To avoid overhead of loading from the disk each particular time series element, 
 * this class group several subsequent time series elements together and store them 
 * as single object (block).
 * </p><p> 
 * As far as Java currently has no templates and
 * Perst need to know format of block class, it is responsibity of prgorammer
 * to create block implementation derived from TimeSeries.Block class
 * and containing array of time series elements. Size of this array specifies
 * the size of the block.
 * </p>
 */
public interface TimeSeries<T extends TimeSeries.Tick> extends IPersistent, IResource, Collection<T> { 
    /**
     * Interface for timeseries element.
     * You should derive your time series element from this class
     * and implement getTime method.
     */
    public interface Tick extends IValue { 
        /**
         * Get time series element timestamp
         * @return timestamp in milliseconds
         */
        long getTime();
    }
    
    /**
     * Abstract base class for time series block.
     * Progammer has to define its own block class derived from this class
     * containign array of time series elements and providing getTicks()
     * method to access this array. It is better no to initialize this array in constructor
     * (because it will be also used when block will be loaded from the disk), 
     * but check in getTicks() method that array is null, and if so - create new array.
     */
    public static abstract class Block extends Persistent { 
        public long timestamp;
        public int  used;

        /**
         * Get time series elements stored in this block.
         * @return preallocated array of time series element. Only <code>used</code>
         * items of this array actually contains time series elements.
         * But all array items should be not null and conain referen to Tick object.
         */
        public abstract Tick[] getTicks();
    }

    /**
     * Add new tick to time series
     * @param tick new time series element
     */
    boolean add(T tick);    

    /**
     * Get list of alements in the time series (in ascending order)
     * @return list of all elements
     */
    ArrayList<T> elements();

    /**
     * Get forward iterator through all time series elements
     * @return forward iterator 
     */
    Iterator<T> iterator();

    /**
     * Get forward iterator for time series elements belonging to the specified range
     * @param from inclusive time of the begging of interval, 
     * if null then take all elements from the beginning of time series
     * @param till inclusive time of the ending of interval, 
     * if null then take all elements till the end of time series
     * @return forward iterator within specified range.
     */
    IterableIterator<T> iterator(Date from, Date till);

    /**
     * Get iterator through all time series elements
     * @param ascent direction of iteration
     * @return iterator in specified direction
     */
    IterableIterator<T> iterator(boolean ascent);

    /**
     * Get forward iterator for time series elements belonging to the specified range
     * @param from inclusive time of the begging of interval, 
     * if null then take all elements from the beginning of time series
     * @param till inclusive time of the ending of interval, 
     * if null then take all elements till the end of time series
     * @param ascent direction of iteration
     * @return iterator within specified range in specified direction
     */
    IterableIterator<T> iterator(Date from, Date till, boolean ascent);

    /**
     * Get timestamp of first time series element
     * @return time of time series start
     */
    Date getFirstTime();

    /**
     * Get timestamp of last time series element
     * @return time of time series end
     */
    Date getLastTime();

    /**
     * Get number of elements in time series
     * @return number of elements in time series
     */
    int size();

    /** 
     * Get tick for specified data
     * @param timestamp time series element timestamp
     * return time series element for the specified timestamp or null
     * if no such element was found
     */
    T getTick(Date timestamp);
    
    /**
     * Check if data is available in time series for the specified time
     * @param timestamp time series element timestamp
     * @return <code>true</code> if there is element in time series with such timestamp, 
     * <code>false</code> otherwise
     */
    boolean has(Date timestamp);

    /**
     * Remove timeseries elements belonging to the specified range
     * @param from inclusive time of the begging of interval, 
     * if null then remove all elements from the beginning of time series
     * @param till inclusive time of the ending of interval, 
     * if null then remove all elements till the end of time series
     * @return number of removed elements
     */
    int remove(Date from, Date till);
}

    